In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

Importing CSV Data as DataFrame

In [3]:
df = pd.read_csv(r'C:\Users\Goodness\Downloads\DataSets\Cars Dataset\vehicles.csv')
data = df.copy()

Show top 5 records

In [4]:
 df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

Show last 5 records

In [5]:
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

Shape of Dataset

In [6]:
df.shape

(426880, 26)

# Dataset Information

The dataset has 426,880 rows and 26 columns. 

The columns are:
* id
* url
* region
* region url
* price	
* year	
* manufacturer	
* model	
* condition	
* cylinders
* fuel
* odometer
* title status
* transmission
* VIN
* Drive
* size	
* type	
* paint_color	
* image_url	
* description	
* county	
* state	
* lat	
* long	
* posting_date


### Checking Missing Values

In [7]:
df.isna().sum()

id                   0
url                  0
region               0
region_url           0
price                0
year              1205
manufacturer     17646
model             5277
condition       174104
cylinders       177678
fuel              3013
odometer          4400
title_status      8242
transmission      2556
VIN             161042
drive           130567
size            306361
type             92858
paint_color     130203
image_url           68
description         70
county          426880
state                0
lat               6549
long              6549
posting_date        68
dtype: int64

### Check duplicates

In [8]:
df.duplicated().sum()

0

There are no duplicated rows in the dataset.

### Check datatypes

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

### Checking the number of unique values in each column

In [10]:
df.nunique()

id              426880
url             426880
region             404
region_url         413
price            15655
year               114
manufacturer        42
model            29667
condition            6
cylinders            8
fuel                 5
odometer        104870
title_status         6
transmission         3
VIN             118264
drive                3
size                 4
type                13
paint_color         12
image_url       241899
description     360911
county               0
state               51
lat              53181
long             53772
posting_date    381536
dtype: int64

### Check dataset stats

In [11]:
df.describe()

id         price           year      odometer  county  \
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05     0.0   
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04     NaN   
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05     NaN   
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00     NaN   
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04     NaN   
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04     NaN   
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05     NaN   
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07     NaN   

                 lat           long  
count  420331.000000  420331.000000  
mean       38.493940     -94.748599  
std         5.841533      18.365462  
min       -84.122245    -159.827728  
25%        34.601900    -111.939847  
50%        39.150100     -88.432600  
75%        42.398900     -80.832039  
max        82.390818     173.885502

The average price of the vehicles above is 75,199 and the average year is 2011. For the odometer, the average is 98,043 miles. The highest price of a vehicle here is 

## Data Exploration

In [12]:
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [13]:


print("Categories in 'manufacturer' variable:     ",end=" " )
print(df['manufacturer'].unique())

print("\nCategories in 'model' variable:     ",end=" " )
print(df['model'].unique())

print("\nCategories in 'condition' variable:     ",end=" " )
print(df['condition'].unique())

print("\nCategories in 'cylinders' variable:     ",end=" " )
print(df['cylinders'].unique())

print("\nCategories in 'fuel' variable:     ",end=" " )
print(df['fuel'].unique())

print("\nCategories in 'odometer' variable:     ",end=" " )
print(df['odometer'].unique())

print("\nCategories in 'title_status' variable:     ",end=" " )
print(df['title_status'].unique())

print("\nCategories in 'transmission' variable:     ",end=" " )
print(df['transmission'].unique())

print("\nCategories in 'drive' variable:     ",end=" " )
print(df['drive'].unique())

print("\nCategories in 'size' variable:     ",end=" " )
print(df['size'].unique())

print("\nCategories in 'type' variable:     ",end=" " )
print(df['type'].unique())

print("\nCategories in 'paint_color' variable:     ",end=" " )
print(df['paint_color'].unique())



Categories in 'manufacturer' variable:      [nan 'gmc' 'chevrolet' 'toyota' 'ford' 'jeep' 'nissan' 'ram' 'mazda'
 'cadillac' 'honda' 'dodge' 'lexus' 'jaguar' 'buick' 'chrysler' 'volvo'
 'audi' 'infiniti' 'lincoln' 'alfa-romeo' 'subaru' 'acura' 'hyundai'
 'mercedes-benz' 'bmw' 'mitsubishi' 'volkswagen' 'porsche' 'kia' 'rover'
 'ferrari' 'mini' 'pontiac' 'fiat' 'tesla' 'saturn' 'mercury'
 'harley-davidson' 'datsun' 'aston-martin' 'land rover' 'morgan']

Categories in 'model' variable:      [nan 'sierra 1500 crew cab slt' 'silverado 1500' ... 'gand wagoneer'
 '96 Suburban' 'Paige Glenbrook Touring']

Categories in 'condition' variable:      [nan 'good' 'excellent' 'fair' 'like new' 'new' 'salvage']

Categories in 'cylinders' variable:      [nan '8 cylinders' '6 cylinders' '4 cylinders' '5 cylinders' 'other'
 '3 cylinders' '10 cylinders' '12 cylinders']

Categories in 'fuel' variable:      [nan 'gas' 'other' 'diesel' 'hybrid' 'electric']

Categories in 'odometer' variable:      [    nan  5

In [14]:
# define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 7 numerical features : ['id', 'price', 'year', 'odometer', 'county', 'lat', 'long']

We have 19 categorical features : ['url', 'region', 'region_url', 'manufacturer', 'model', 'condition', 'cylinders', 'fuel', 'title_status', 'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color', 'image_url', 'description', 'state', 'posting_date']


There are 7 numerical features and 19 categorical features in the dataset.


In [15]:
##No of Cars below and above the average price
above_ave  = df[df['price'] > 75199]['price'].count()
below_ave = df[df['price'] < 75199]['price'].count()

print(f'Number of cars that got sold above the average price is: {above_ave}')
print(f'Number of cars that got sold below the average price is: {below_ave}')

Number of cars that got sold above the average price is: 2296
Number of cars that got sold below the average price is: 424584


In [16]:
##No of Cars below and above the average odometer mileage
above_ave_mile  = df[df['price'] > 98043]['odometer'].count()
below_ave_mile = df[df['price'] < 98043]['odometer'].count()

print(f'Number of cars that got sold whose mileage is above the average mileage is: {above_ave_mile}')
print(f'Number of cars that got sold whose mileage is below the average mileage is {below_ave_mile}')

Number of cars that got sold whose mileage is above the average mileage is: 762
Number of cars that got sold whose mileage is below the average mileage is 421718


### Insight

The greater number of cars sold were below the average price, while just about 2296 cars were sold above the average price in the dataset. Also, the greater number of cars had not been used above the average mileage of 98043.

### Handling Missing Values In the dataset

Handling the missing values in the dataset and seeing what their new stats are


In [18]:
# Drop columns with more than 50% missing values
threshold = len(df) * 0.5
df = df.dropna(thresh=threshold, axis=1)

In [19]:
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ...  VIN  drive type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN    NaN  NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN    NaN  NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN    NaN  NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN    NaN  NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN    NaN  NaN         NaN   

  image_url description state lat long posting_date  
0       NaN         NaN    az NaN  NaN          NaN  
1       NaN         NaN    ar NaN  NaN          NaN  
2       NaN         NaN    fl NaN  NaN          NaN  
3       NaN         NaN    ma NaN  NaN          NaN  
4       NaN         NaN    nc NaN  NaN          NaN  

[5 rows x 24 columns]

In [20]:
# Drop rows with missing values in crucial columns
crucial_columns = ['price', 'year', 'manufacturer', 'model', 'odometer', 'title_status', 'transmission']
df = df.dropna(subset=crucial_columns)

In [21]:
# Impute missing values for numerical columns
numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns
df[numerical_cols] = df[numerical_cols].fillna(df[numerical_cols].mean())

In [23]:
# Impute missing values for categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns
df[categorical_cols] = df[categorical_cols].fillna(df[categorical_cols].mode().iloc[0])

In [24]:
df.shape

(391154, 24)

In [25]:
# Save the cleaned dataset
df.to_csv('cleaned_dataset.csv', index=False)

In [26]:
df_subset = df.sample(frac=0.08)
#Saving subset of cleaned dataset
df_subset.to_csv('cleaned_dataset.csv', index=False)